panic: send on closed channel

Repro is with Zed commit 2b3d086. The test data is too large to attach in GitHub, so I've uploaded it to a public Google Cloud storage bucket. It's just ZNG generated after importing the well-known "all.pcap" into Brim.

To prepare, I initialize a new Zed Lake and import the test data.

```
$ zed -version
Version: v0.29.0-394-g2b3d086f

$ rm -rf $ZED_LAKE_ROOT

$ zed lake init
using environment variable ZED_LAKE_ROOT
lake created: /Users/phil/logs

$ zed lake create -p logs
pool created: logs

$ zed lake load -p logs https://storage.googleapis.com/brimsec-public-test-data/zed-2822/all.zng
1uJbPg65ZFKxImTqxNVh0U4LlWF committed
```

At this point, I execute the following query and then quickly interrupt it via Ctrl-C, repeating this several times in a row. Eventually it crashes with the stack dump shown:

```
$ zed lake query -z 'from logs | put id.orig_p:=uint16(id.orig_p),id.resp_p:=uint16(id.resp_p),id.proto:=string(proto) | sort id | count()'
...
panic: send on closed channel

goroutine 71 [running]:
github.com/brimdata/zed/proc/put.(*Proc).maybeWarn(0xc000434b40, 0x1deef80, 0xc000118670)
	/Users/phil/work/zed/proc/put/put.go:75 +0xda
github.com/brimdata/zed/proc/put.(*Proc).put(0xc000434b40, 0xc000301080, 0xc000301080, 0x0, 0x0)
	/Users/phil/work/zed/proc/put/put.go:348 +0x3d5
github.com/brimdata/zed/proc/put.(*Proc).Pull(0xc000434b40, 0xc005cf1a40, 0xc005cf1a40, 0x0, 0x0)
	/Users/phil/work/zed/proc/put/put.go:371 +0x165
github.com/brimdata/zed/proc/sort.(*Proc).recordsForOneRun(0xc0003ac720, 0x0, 0x0, 0x0, 0x0, 0xc000067fc8, 0x1450a56)
	/Users/phil/work/zed/proc/sort/sort.go:103 +0x1ca
github.com/brimdata/zed/proc/sort.(*Proc).sortLoop(0xc0003ac720)
	/Users/phil/work/zed/proc/sort/sort.go:59 +0x78
created by github.com/brimdata/zed/proc/sort.(*Proc).Pull.func1
	/Users/phil/work/zed/proc/sort/sort.go:46 +0x3e
```

The attached video shows me up-arrowing and repeating it several times, just to give you a feel for the timing. It's definitely not a "it works on the first try" kind of repro.

https://user-images.githubusercontent.com/5934157/122980723-da19d280-d34d-11eb-8ac2-45b308927784.mp4

What's also troubling is that I initially was attempting this query using `zapi` against the Pool that had been created by importing the pcap into Brim. The first way I learned something had gone wrong was when I went to re-run my `zapi` and saw the error message:

```
Post "http://localhost:9867/query": dial tcp [::1]:9867: connect: connection refused
```

I looked into both the `main.log` and the `zlake.log` associated with the running app, but there was no crash message. That's when I went to doing the repro at the Lake level outside the app as shown above. This doesn't bode well for field-debug-ability if crashes don't make it into the log, so a separate issue may be justified here. I'll wait for a Dev to provide guidance as to whether this part is considered a Brim issue or a Zed issue.

As proof, the attached video shows me doing that repro while running Brim commit `69334c0` which uses Zed commit 2b3d086 as its backend. I'm tailing both the `main.log` and `zlake.log` and you can see the crash message never appears.

https://user-images.githubusercontent.com/5934157/122982384-a8097000-d34f-11eb-815e-70211d2be9ad.mp4

--- a/zbuf/merger.go
+++ b/zbuf/merger.go
@@ -24,6 +24,7 @@ type Merger struct {
 	cmp     expr.CompareFn
 	once    sync.Once
 	pullers []*mergerPuller
+	wg      sync.WaitGroup
 }
 
 type mergerPuller struct {
@@ -106,8 +107,10 @@ func MergeByTs(ctx context.Context, pullers []Puller, o order.Which) *Merger {
 
 func (m *Merger) run() {
 	for _, p := range m.pullers {
+		m.wg.Add(1)
 		puller := p
 		go func() {
+			defer m.wg.Done()
 			for {
 				b, err := puller.Pull()
 				select {
@@ -131,6 +134,7 @@ func (m *Merger) Read() (*zng.Record, error) {
 	leader, err := m.findLeader()
 	if leader < 0 || err != nil {
 		m.Cancel()
+		m.wg.Wait()
 		return nil, err
 	}
 	return m.pullers[leader].next(), nil
@@ -198,6 +202,7 @@ func (m *Merger) Pull() (Batch, error) {
 	leader, err := m.findLeader()
 	if leader < 0 || err != nil {
 		m.Cancel()
+		m.wg.Wait()
 		return nil, err
 	}
 	if !m.overlaps(leader) {

